functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,168 @@
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("`data` should not be empty")
+		}
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, legacyFnConfigKind):
+		fallthrough
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind):
+		p.NewLabels = o.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+	}
+	// add default fields
+	if err := p.addDefaultLabelFields(); err != nil {
+		return err
+	}
+	// add additional fields
+	if o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind) {
+		arr, exist, err := o.NestedSlice("additionalLabelFields")
+		if err != nil {
+			return fmt.Errorf("`additionalLabelFields` format is wrong")
+		}
+		if exist {
+			var addFields []FieldSpec
+			for _, sub := range arr {
+				var curField FieldSpec
+				if err := sub.As(&curField); err != nil {
+					return err
+				}
+				addFields = append(addFields, curField)
+			}
+			p.AdditionalLabelFields = append(p.AdditionalLabelFields, addFields...)
+		}
+	}
+	return nil
+}
+
+func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+	for _, o := range objects {
+		for _, sp := range p.AdditionalLabelFields {
+			if (sp.Group == "" && sp.Version == "" && sp.Kind == "") || o.IsGVK(sp.Group, sp.Version, sp.Kind) {
+				newResult := fn.Result{
+					Message:     "Replace labels",
+					Severity:    "INFO",
+					ResourceRef: nil, // TODO: initialize in kyaml
+					Field: &fn.Field{
+						Path:          sp.Path,
+						CurrentValue:  nil, // values to be updated in setLabel()
+						ProposedValue: nil, // values to be updated in setLabel()
+					},
+					File: &fn.File{
+						Path:  o.PathAnnotation(),
+						Index: o.IndexAnnotation(),
+					},
+					Tags: nil,
+				}
+				err := updateLabels(o, sp.Path, p.NewLabels, sp.CreateIfNotPresent, &newResult)
+				if err != nil {
+					return err
+				}
+				p.Results = append(p.Results, &newResult)
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: Is there any helper function insdie fn package to help with copy map
+func copyStringMap(m map[string]string) map[string]string {
+	if m == nil {
+		return nil
+	}
+	newMap := make(map[string]string)
+	for k, v := range m {
+		newMap[k] = v
+	}
+	return newMap
+}
+
+func updateLabels(o *fn.KubeObject, fieldPath string, newLabels map[string]string, create bool, result *fn.Result) error {
+	paths := strings.Split(fieldPath, "/")
+
+	oldLabels, exist, err := o.NestedStringMap(paths...)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		oldLabels = make(map[string]string)
+	}
+	result.Field.CurrentValue = copyStringMap(oldLabels)
+	replaceLabels(oldLabels, newLabels, create)
+	result.Field.ProposedValue = copyStringMap(oldLabels)
+	o.SetNestedStringMapOrDie(oldLabels, paths...)
```

Comment:
This makes it looks like we need more result related functions in the SDK. 
I'd suggest 1. drop the resource level result to unblock this PR 2. add result methods in the Go SDK to make this kinds of `copyStringMap` easier. 3 support https://github.com/GoogleContainerTools/kpt/issues/2996 to display results by severity level 4. add back the resource level result.